### PR TITLE
[SPARK-17548] [MLlib] Word2VecModel.findSynonyms no longer spuriously rejects the best match when invoked with a vector

### PR DESCRIPTION
**Pull request by willb**
## What changes were proposed in this pull request?

This pull request changes the behavior of `Word2VecModel.findSynonyms` so that it will not spuriously reject the best match when invoked with a vector that does not correspond to a word in the model's vocabulary.  Instead of blindly discarding the best match, the changed implementation discards a match that corresponds to the query word (in cases where `findSynonyms` is invoked with a word) or that has an identical angle to the query vector.
## How was this patch tested?

I added a test to `Word2VecSuite` to ensure that the word with the most similar vector from a supplied vector would not be spuriously rejected.

### CODE REVIEW

**Comment by hhbyyh** [comment]
Hi @willb Good catch. This is a valid issue.

**Comment by willb** [comment]
Thanks, @hhbyyh.  This is what I get for running `sbt "testOnly ..."`  I'll push a fix.

File: mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala [hunk 1]
```diff
@@ -566,8 +582,8 @@ class Word2VecModel private[spark] (
     wordList.zip(cosVec)
       .toSeq
       .sortBy(-_._2)
-      .take(num + 1)
-      .tail
+      .filter(tup => wordOpt.map(w => !w.equals(tup._1)).getOrElse(true) && tup._2 != 1.0d)
```
Comment (hhbyyh) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala]: Due to the floating point calculation error, `tup._2` may not be `1.0d` even for the same vector. 
And I 'm not sure if it's alway proper to reject the identical vector.

File: mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala [hunk 2]
```diff
@@ -566,8 +582,8 @@ class Word2VecModel private[spark] (
...
+      .take(num)
```
Comment (hhbyyh) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala]: Just for performance:
`.take(num + 1)`
`.filter(...)`
may be faster.

File: mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala (hunk 1 above)
Comment (willb) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala]: Yes, those are both valid points.  I went back and forth on this one, but I think we could actually argue that rejecting the identical vector doesn't make sense in any case.

File: mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala (hunk 2 above)
Comment (willb) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala]: We'd actually need `.take(num + 1).filter(...).take(num)` to properly handle the cases where the filter isn't rejecting anything.  I'm assuming the filter is fairly cheap but you're right that it doesn't make sense to do it any more than is necessary.

File: mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala (hunk 1 above)
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala]: It's probably easier still to leave the current code in place, up to "tail". If wordOpt is defined, then apply filter and take(num). If not, apply tail.

File: mllib/src/test/scala/org/apache/spark/mllib/feature/Word2VecSuite.scala [hunk 3]
```diff
@@ -68,6 +69,21 @@ class Word2VecSuite extends SparkFunSuite with MLlibTestSparkContext {
...
+  test("findSynonyms doesn't reject similar word vectors when called with a vector") {
+    val num = 2
+    val word2VecMap = Map(
+      ("china", Array(0.50f, 0.50f, 0.50f, 0.50f)),
+      ("japan", Array(0.40f, 0.50f, 0.50f, 0.50f)),
+      ("taiwan", Array(0.60f, 0.50f, 0.50f, 0.50f)),
+      ("korea", Array(0.45f, 0.60f, 0.60f, 0.60f))
+    )
+    val model = new Word2VecModel(word2VecMap)
+    val syms = model.findSynonyms(Vectors.dense(Array(0.52d, 0.50d, 0.50d, 0.50d)), num)
```
Comment (srowen) [inline mllib/src/test/scala/org/apache/spark/mllib/feature/Word2VecSuite.scala]: Nit, but I think "d" is redundant. Also "0.5" seems clearer than "0.50" but this is truly up to taste.

**Review by srowen — CHANGES_REQUESTED** [review]

File: mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala (hunk 1 above)
Comment (willb) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala]: @hhbyyh So [one of the python doctests](https://github.com/apache/spark/blob/branch-2.0/python/pyspark/mllib/feature.py#L575) depends on a word not being a synonym of its vector representation.  I think since this is the documented behavior now, that's the direction the fix should go as well, but I'll use `Vector.equals` instead of checking similarity in any case.

Comment (willb) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala]: @srowen This code in general kind of bothers me (I'd rather see a single pass through the tuples with a bounded priority queue keeping track of the `num + 1` candidates than converting to a sequence and then allocating an array to sort in place).  But I'm inclined to get some numbers showing that that is a good idea and make it a separate PR unless this is a good time to fold it in (so to speak).

File: mllib/src/test/scala/org/apache/spark/mllib/feature/Word2VecSuite.scala (hunk 3 above)
Comment (willb) [inline mllib/src/test/scala/org/apache/spark/mllib/feature/Word2VecSuite.scala]: Yes to both (the two significant digits was simply following the style earlier in the test).

File: mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala (hunk 1 above)
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala]: Yeah it's not efficient, and there's a todo about it above. I think you could treat it as a separate issue. I'm pretty certain it's a win.

Comment (willb) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala]: I wasn't sure if the todo was referring to the sorting or to earlier optimizations.  I'll get it started as a separate issue; thanks!

Comment (willb) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala]: @srowen So actually reverting to the old code but filtering only if `wordOpt` is defined doesn't handle the original case I was considering here, where you pass in a vector that is very similar to the representation of a word in the vocabulary but that is not itself the representation of a word in the vocabulary.

Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala]: It should be equivalent to what you suggest. Something like...

```
val topn1 = wordList.zip(cosVec).toSeq.sortBy(-_._2).take(num + 1)
if (wordOpt.isDefined) {
  topn1.filter(tup => !wordOpt.get.equals(tup._1)).take(num)
} else {
  topn1.tail
}
```

OK that's a bit different than what you suggested but does that help a bit?

**Comment by willb** [comment]
Can we assume in general that distinct words will not have identical (viz., by `.equals`) vector representations?  I ask because the behavior in the current documentation (in that pyspark test I linked above) assumes that you don't want a word to turn up as a synonym for its own vector representation.  But this is impossible to enforce if we aren't guaranteed that distinct words won't have identical vector representations.

This seems like a mostly reasonable assumption but it is definitely a corner case that we might want to be robust to (or at least take note of).  If we can't accept this assumption, then we should figure out whether or not changing the expected behavior in the documentation is acceptable.

**Comment by srowen** [comment]
I don't think you can assume that, though it will almost always be true. But I don't think the goal is to filter vectors of other words that happen to be identical. The point was just to filter the word itself from the list of synonyms because it's always perfectly similar to itself and that's implicitly not what the caller wants.

So, no you don't necessarily want to filter out all "1.0" similarity.

File: mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala (hunk 1 above)
Comment (willb) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala]: Consider the case where you're passing in a vector that is extremely similar to the representation of a word in the vocabulary but that is not itself the representation of a word in the vocabulary.  (A concrete example is (in this test I added)[https://github.com/willb/spark/blob/fix/findSynonyms/mllib/src/test/scala/org/apache/spark/mllib/feature/Word2VecSuite.scala#L72].)  In this case, `wordOpt` is not defined (because you are querying for words whose vector representations are similar to a given vector) but you nonetheless are not interested in discarding the best match because it is not a trivial match (that is, it is not going to be your query term).

Related to your other comment (and discussion elsewhere on this PR), I think we could make a case for changing the documented behavior (especially since it is only documented as such in pyspark) in the case where `findSynonyms` is invoked with the vector representation of a word that is _in_ the vocabulary.  Instead of rejecting the best match in that case, we could return it. The argument there is that such a result is telling you something you didn't necessarily know (otherwise, you'd probably be querying for a word and not a vector) and that there is an easy way to identify that such a match is trivially the best match.  I recognize that changing documented behavior is a big deal, but in this case it seems like it could be the best way to address a minor bug.

Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala]: Oops, "tail" is wrong, yeah, because it would give you the bottom n out of n+1, when you just want the first n out of the n+1. Otherwise I think this works?

Anyway, agree with the change you propose. If specifying a vector I would not expect any filtering of the first element. We're changing the behavior no matter what here but it's a fix.

**Comment by willb** [comment]
@srowen Yes.  But if we're querying for words similar to a given vector, then there's no word to filter out (and, indeed, no way to know which word we might want to filter out if multiple words might map to the same vector representation).

File: mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala [hunk 4]
```diff
@@ -563,11 +580,16 @@ class Word2VecModel private[spark] (
       ind += 1
     }
 
+    // NB: This code (and the documented behavior of findSynonyms
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala]: Oh, is the problem that several words, including the word itself, may have 1.0 similarity? and so the word itself may not sort as the top result among the 1.0 results? Yeah, then the code below doesn't quite handle that case. (But below you wouldn't need the `take(num + 1)` now I think?)

What about just ...

```
val scored = wordList.zip(cosVec).toSeq.sortBy(-_._2)
val filtered = wordOpt match {
  Some(word) => scored.take(num + 1).filter(p => word != p._1)
  None => scored
}
filtered.take(num).toArray
```

**Review by srowen — CHANGES_REQUESTED** [review]

File: mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala [hunk 5]
```diff
@@ -227,7 +227,7 @@ class Word2VecModel private[ml] (
    */
   @Since("1.5.0")
   def findSynonyms(word: String, num: Int): DataFrame = {
-    findSynonyms(wordVectors.transform(word), num)
+    findSynonyms(wordVectors.transform(word), num, Some(word))
```
Comment (willb) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala]: In this case (and similarly in [`Word2VecModelWrapper`](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/mllib/api/python/Word2VecModelWrapper.scala)) I opted to call the three-argument version because the wrappers both explicitly convert their argument to a vector before calling `findSynonyms` on the underlying model (and so `wordOpt` would not be defined if the wrapper were invoked with a word).  If we were to make the three-argument `findSynonyms` private we wouldn't be able to share a code path in the wrapper classes and would need to duplicate the code to tidy and reformat results in both methods (data frame creation in this case, unzipping and `asJava` in the Python model wrapper) or factor it out to a separate method.  Let me know how you want me to proceed here.

I agree that updating the docs makes sense and will make it clearer to future maintainers as well.

Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala]: I see. I agree that something gets a little bit duplicated no matter what. Given your change, it seems easiest to pass the string vs vector argument all the way down, even if that means in these other two classes you duplicate a little code to transform the dataframe, etc. If it's nontrivial, sure, a little helper method seems reasonable. That would help keep this layered more cleanly IMHO.

**Comment by willb** [comment]
Thanks for the feedback @srowen!  I think 18e6bfe addresses everything from a code perspective, but it missed removing the comment about assuming that distinct words have distinct vector representations, so I've just pushed another commit that just removes that comment.

**Review by srowen — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master/2.0
